src/toil/lib/docker.py

```diff
@@ -1,280 +1,361 @@
-            else:
-                containerName = baseDockerCall[baseDockerCall.index('--name') + 1]
-        else:
-            containerName = _getContainerName(job)
-            baseDockerCall.extend(['--name', containerName])
-    except ValueError:
-        containerName = _getContainerName(job)
-        baseDockerCall.extend(['--name', containerName])
-    except IndexError:
-        raise RuntimeError("Couldn't parse Docker's `--name=` option, check parameters: " + str(dockerParameters))
-
-    # Defer the container on-exit action
-    if '--rm' in baseDockerCall and defer is None:
-        defer = RM
-    if '--rm' in baseDockerCall and defer is not RM:
-        _logger.warn('--rm being passed to docker call but defer not set to dockerCall.RM, defer set to: ' + str(defer))
-    job.defer(_dockerKill, containerName, action=defer)
-    # Defer the permission fixing function which will run after this job concludes.
-    # We call this explicitly later on in this function, but we defer it as well to handle unexpected job failure.
-    job.defer(_fixPermissions, tool, workDir)
-
-    # Make subprocess call
-
-    # If parameters is list of lists, treat each list as separate command and chain with pipes
+    # If 'parameters' is a list of lists, treat each list as a separate command
+    # and chain with pipes.
     if len(parameters) > 0 and type(parameters[0]) is list:
-        # When piping, all arguments now get merged into a single string to bash -c.
-        # We try to support spaces in paths by wrapping them all in quotes first.
-        chain_params = [' '.join(p) for p in [list(map(pipes.quote, q)) for q in parameters]]
-        # Use bash's set -eo pipefail to detect and abort on a failure in any command in the chain
-        call = baseDockerCall + ['--entrypoint', '/bin/bash',  tool, '-c',
-                                 'set -eo pipefail && {}'.format(' | '.join(chain_params))]
-    else:
-        call = baseDockerCall + [tool] + parameters
-    _logger.info("Calling docker with " + repr(call))
-
-    params = {}
-    if outfile:
-        params['stdout'] = outfile
-    if checkOutput:
-        callMethod = subprocess.check_output
+        chain_params = \
+            [' '.join((pipes.quote(arg) for arg in command)) \
+             for command in parameters]
+        command = ' | '.join(chain_params)
+        pipe_prefix = "set -eo pipefail && "
+        command = [pipe_prefix + command]
+        _logger.debug("Calling docker with: " + repr(command))
+
+    # If 'parameters' is a normal list, join all elements into a single string
+    # element.
+    # Example: ['echo','the', 'Oread'] becomes: ['echo the Oread']
+    # Note that this is still a list, and the docker API prefers this as best
+    # practice:
+    # http://docker-py.readthedocs.io/en/stable/containers.html
+    elif len(parameters) > 0 and type(parameters) is list:
+        command = [' '.join(p) for p in parameters]
```

Comment:
`command = ' '.join(parameters)` is what you want here, otherwise you put a space between every character.

Reply:
Updated.